hivemq-edge/src/main/java/com/hivemq/edge/modules/adapters/impl/polling/PollingTask.java

```diff
@@ -0,0 +1,210 @@
+        if (stopBecauseOfTooManyErrors) {
+            log.warn(
+                    "Detected bad system process {} in sampler {} - terminating process to maintain health ({}ms runtime)",
+                    errorCountTotal,
+                    sampler.getAdapterId(),
+                    milliSecondsSinceLastPoll);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug(
+                        "Detected bad system process {} in sampler {} - interrupted process to maintain health ({}ms runtime)",
+                        errorCountTotal,
+                        sampler.getAdapterId(),
+                        milliSecondsSinceLastPoll);
+            }
+        }
+        notifyOnError(sampler, throwable, !stopBecauseOfTooManyErrors);
+        if (!stopBecauseOfTooManyErrors) {
+            reschedule(errorCountTotal);
+        }
+    }
+
+
+    private void handleExceptionDuringPolling(final @NotNull Throwable throwable) {
+        int errorCountTotal = applicationErrorCount.incrementAndGet();
+        final int maxErrorsBeforeRemoval = sampler.getMaxErrorsBeforeRemoval();
+        // case 1: Unlimited retry (maxErrorsBeforeRemoval < 0) or less errors than the limit
+        if (maxErrorsBeforeRemoval < 0 || errorCountTotal <= maxErrorsBeforeRemoval) {
+            if (log.isDebugEnabled()) {
+                log.debug("Application Error {} in sampler {} -> {}",
+                        errorCountTotal,
+                        sampler.getAdapterId(),
+                        throwable.getMessage(),
+                        throwable);
+            }
+            reschedule(errorCountTotal);
+            notifyOnError(sampler, throwable, true);
+        } else {
+            log.info(
+                    "Detected '{}' recent errors when sampling. This exceeds the configured limit of '{}'. Sampling for adapter with id '{}' gets stopped.",
+                    errorCountTotal,
+                    maxErrorsBeforeRemoval,
+                    sampler.getAdapterId());
+            notifyOnError(sampler, throwable, false);
+            // no rescheduling
+        }
+
+    }
+
+    private void notifyOnError(
+            final @NotNull ProtocolAdapterPollingSampler sampler, final @NotNull Throwable t, boolean continuing) {
+        try {
+            sampler.error(t, continuing);
+        } catch (Throwable samplerError) {
+            if (log.isInfoEnabled()) {
+                log.info("Sampler Encountered Error In Notification", samplerError);
+            }
+        }
+    }
+
+    private void reschedule(int errorCountTotal) {
+        final long delayInMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - nanosOfLastPolling);
+        // a negative delay means that the last polling attempt took longer to be processed than the specified delay between polls
+        if (delayInMillis < 0) {
```

Comment:
I don't see how the condition could every be true: `System.nanoTime()` will always be >= `nanosOfLastPolling`. You probably wanted to compare it to `sampler.getPeriod()`?

Reply:
yeah that is missing for sure